File: leetcode/leetcode_p474/code_474.py
# 力扣第474题，1和0
"""
在计算机界中，我们总是追求用有限的资源获取最大的收益。

现在，假设你分别支配着 m 个 0 和 n 个 1。另外，还有一个仅包含 0 和 1 字符串的数组。

你的任务是使用给定的 m 个 0 和 n 个 1 ，找到能拼出存在于数组中的字符串的最大数量。每个 0 和 1 至多被使用一次。

注意:

给定 0 和 1 的数量都不会超过 100。
给定字符串数组的长度不会超过 600。
"""



def counts(strs):
    num_0 = strs.count('0')
    num_1 = strs.count('1')
    return [num_0, num_1]


def findMaxForm(strs, m, n):
    dp = [[0]*(n+1) for _ in range(m+1)]
    max_num = 0
    for s in strs:
        mn = counts(s)
        for mi in range(m, -1, -1):
            if mi - mn[0] < 0:
                break
            for ni in range(n, -1, -1):
                if ni - mn[1] < 0:
                    break
                dp[mi][ni] = max(dp[mi-mn[0]][ni-mn[1]]+1, dp[mi][ni])
                # print(dp)
                if dp[mi][ni] > max_num:
                    max_num = dp[mi][ni]
    #     print(dp)
    print(dp)
    return max_num


if __name__ == "__main__":
    strs = ['10', '0001', '111001', '1', '0']
    m = 5 
    n = 3
    # strs = ["10","0","1"]
    # m = 1
    # n = 1
    # for s in strs:
    #     print(counts(s))
    result = findMaxForm(strs, m, n)
    print('result:', result)

File: leetcode/leetcode_p474/code_474.cpp
// 力扣第474题，1和0
/*
在计算机界中，我们总是追求用有限的资源获取最大的收益。

现在，假设你分别支配着 m 个 0 和 n 个 1。另外，还有一个仅包含 0 和 1 字符串的数组。

你的任务是使用给定的 m 个 0 和 n 个 1 ，找到能拼出存在于数组中的字符串的最大数量。每个 0 和 1 至多被使用一次。

注意:

给定 0 和 1 的数量都不会超过 100。
给定字符串数组的长度不会超过 600。
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int findMaxForm(vector<string>& strs, int m, int n){
        vector<vector<int>> dp(m+1, vector<int>(n+1, 0));

        int max_value = 0;
        for (int s = 0; s < strs.size(); ++s){
            int *mn = count_01(strs[s]);
            cout << "mn[0]: " << mn[0] << "mn[1]: " << mn[1] << endl;
            for (int mi = m; mi >= 0; --mi){
                if (mi - mn[0] < 0)
                    break;
                for (int ni = n; ni >= 0; --ni){
                    if (ni - mn[1] < 0)
                        break;
                    // cout << "dp[mi-mn[0]][ni-mn[1]]+1: " << dp[mi-mn[0]][ni-mn[1]]+1 << endl;
                    dp[mi][ni] = max(dp[mi-mn[0]][ni-mn[1]]+1, dp[mi][ni]);
                    if (dp[mi][ni] > max_value)
                        max_value = dp[mi][ni];
                }
            }
            delete [] mn;
        }

        return max_value;
    }

    int *count_01(string s){
        int num0 = count(s.begin(), s.end(), '0');
        int num1 = count(s.begin(), s.end(), '1');
        int * num01 = new int[2]{num0, num1};
        return num01;
    }
};


int main()
{
    string s = "00011";
    int num0 = count(s.begin(), s.end(), '0');
    int num1 = count(s.begin(), s.end(), '1');
    int num2 = count(s.begin(), s.end(), '2');
    cout << "num0 = " << num0 << endl;
    cout << "num1 = " << num1 << endl;
    cout << "num2 = " << num2 << endl;

    vector<string> strs = {"10", "0001", "111001", "1", "0"};
    int m = 5;
    int n = 3;
    Solution sl;
    int result = sl.findMaxForm(strs, m, n);
    cout << "result: " << result << endl;

    cin.get();
    return 0;
}
